File: macos/Sources/Transkrito/Dictionary/DictionaryStore.swift
import Foundation
import Observation

/// Owns dictionary.json. UI edits write atomically; external edits (TextEdit, sync) reload via a DispatchSource
/// watching the directory. A malformed file keeps the last good in-memory copy and sets `loadError`; the file is
/// not overwritten until the user saves from the UI.
@MainActor @Observable
final class DictionaryStore {
    private(set) var entries: [DictionaryEntry] = []
    private(set) var loadError: String?
    var onChange: (() -> Void)?

    let fileURL: URL
    private var lastWrite: Date?
    private var watcher: DispatchSourceFileSystemObject?
    private var dirFD: Int32 = -1

    init(fileURL: URL = AppPaths.dictionary, watch: Bool = true) {
        self.fileURL = fileURL
        try? FileManager.default.createDirectory(at: fileURL.deletingLastPathComponent(), withIntermediateDirectories: true)
        load()
        if !FileManager.default.fileExists(atPath: fileURL.path) { save() }
        if watch { startWatching() }
    }

    private func startWatching() {
        let dir = fileURL.deletingLastPathComponent()
        dirFD = open(dir.path, O_EVTONLY)
        guard dirFD >= 0 else { return }
        let src = DispatchSource.makeFileSystemObjectSource(fileDescriptor: dirFD, eventMask: [.write, .rename, .extend], queue: .main)
        src.setEventHandler { [weak self] in Task { @MainActor in self?.externalChange() } }
        src.setCancelHandler { [dirFD] in close(dirFD) }
        src.resume()
        watcher = src
    }

    private func externalChange() {
        // Debounce editors that write twice, and ignore the echo of our own atomic write.
        Task { @MainActor [weak self] in
            try? await Task.sleep(for: .milliseconds(150))
            guard let self else { return }
            let stamp = (try? FileManager.default.attributesOfItem(atPath: self.fileURL.path)[.modificationDate]) as? Date
            if stamp == self.lastWrite { return }
            self.lastWrite = stamp
            self.load()
            self.onChange?()
        }
    }

    private func load() {
        do {
            let file = try JsonFile.read(DictionaryFile.self, from: fileURL) ?? DictionaryFile()
            entries = file.entries
            loadError = nil
            lastWrite = (try? FileManager.default.attributesOfItem(atPath: fileURL.path)[.modificationDate]) as? Date
        } catch {
            loadError = "dictionary.json could not be read: \(error.localizedDescription)"
        }
    }

    func save() {
        do {
            try JsonFile.write(DictionaryFile(entries: entries), to: fileURL)
            lastWrite = (try? FileManager.default.attributesOfItem(atPath: fileURL.path)[.modificationDate]) as? Date
            loadError = nil
        } catch {
            loadError = "dictionary.json could not be written: \(error.localizedDescription)"
        }
        onChange?()
    }

    func add(_ entry: DictionaryEntry) { entries.insert(entry, at: 0); save() }
    func remove(_ entry: DictionaryEntry) { entries.removeAll { $0.id == entry.id }; save() }
    func update(_ entry: DictionaryEntry) {
        if let i = entries.firstIndex(where: { $0.id == entry.id }) { entries[i] = entry }
        save()
    }

    func search(_ query: String) -> [DictionaryEntry] {
        let q = query.trimmingCharacters(in: .whitespaces)
        if q.isEmpty { return entries }
        return entries.filter { $0.hearText.localizedCaseInsensitiveContains(q) || $0.writeText.localizedCaseInsensitiveContains(q) }
    }
}

File: windows/Transkrito/Dictionary/DictionaryStore.cs
using System.Collections.ObjectModel;
using Transkrito.Storage;

namespace Transkrito.Dictionary;

/// <summary>
/// Owns dictionary.json. UI edits write atomically; external edits (Notepad, sync) reload via FileSystemWatcher.
/// A malformed file keeps the last good in-memory copy and sets <see cref="LoadError"/>; the file is not
/// overwritten until the user saves from the UI.
/// </summary>
public sealed class DictionaryStore : IDisposable
{
    private readonly string _path;
    private readonly FileSystemWatcher? _watcher;
    private readonly SynchronizationContext? _ui;
    private DateTime _lastWrite;
    private readonly object _gate = new();

    public ObservableCollection<DictionaryEntry> Entries { get; } = new();
    public string? LoadError { get; private set; }
    public event Action? Changed;

    public DictionaryStore(string? path = null, bool watch = true)
    {
        _path = path ?? AppPaths.Dictionary;
        _ui = SynchronizationContext.Current;
        Directory.CreateDirectory(Path.GetDirectoryName(_path)!);
        Load();
        if (!File.Exists(_path)) Save();

        if (!watch) return;
        _watcher = new FileSystemWatcher(Path.GetDirectoryName(_path)!, Path.GetFileName(_path))
        {
            NotifyFilter = NotifyFilters.LastWrite | NotifyFilters.FileName | NotifyFilters.Size,
            EnableRaisingEvents = true,
        };
        _watcher.Changed += OnExternalChange;
        _watcher.Created += OnExternalChange;
        _watcher.Renamed += OnExternalChange;
    }

    public string FilePath => _path;

    private void OnExternalChange(object sender, FileSystemEventArgs e)
    {
        // Ignore the echo of our own atomic write; debounce editors that write twice.
        Task.Delay(150).ContinueWith(_ =>
        {
            lock (_gate)
            {
                if (!File.Exists(_path)) return;
                var stamp = File.GetLastWriteTimeUtc(_path);
                if (stamp == _lastWrite) return;
                _lastWrite = stamp;
            }
            Post(() => { Load(); Changed?.Invoke(); });
        });
    }

    private void Post(Action a)
    {
        if (_ui is null) a(); else _ui.Post(_ => a(), null);
    }

    private void Load()
    {
        try
        {
            var file = JsonFile.Read<DictionaryFile>(_path) ?? new DictionaryFile();
            Entries.Clear();
            foreach (var e in file.Entries) Entries.Add(e);
            LoadError = null;
            if (File.Exists(_path)) _lastWrite = File.GetLastWriteTimeUtc(_path);
        }
        catch (Exception ex)
        {
            LoadError = $"dictionary.json could not be read: {ex.Message}";
        }
    }

    public void Save()
    {
        lock (_gate)
        {
            JsonFile.Write(_path, new DictionaryFile { Entries = Entries.ToList() });
            _lastWrite = File.GetLastWriteTimeUtc(_path);
            LoadError = null;
        }
        Changed?.Invoke();
    }

    public void Add(DictionaryEntry entry) { Entries.Insert(0, entry); Save(); }
    public void Remove(DictionaryEntry entry) { Entries.Remove(entry); Save(); }
    public void Update() => Save();

    public IEnumerable<DictionaryEntry> Search(string query)
    {
        if (string.IsNullOrWhiteSpace(query)) return Entries;
        return Entries.Where(e =>
            e.HearText.Contains(query, StringComparison.OrdinalIgnoreCase) ||
            e.WriteText.Contains(query, StringComparison.OrdinalIgnoreCase));
    }

    public void Dispose() => _watcher?.Dispose();
}
